### PR TITLE
Deprecate hex2num and num2hex

### PR DESCRIPTION
**Pull request by simonbyrne**
See #22031. I'm open to suggestions for the deprecation message for hex2num.

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -1342,6 +1342,22 @@ end
...
+    end
+    if length(s) <= 8
+        return reinterpret(Float32, parse(UInt32, s, 16))
+    end
+    return reinterpret(Float64, parse(UInt64, s, 16))
+end
+
+@deprecate num2hex(x::Union{Float16,Float32,Float64}) =
+    hex(reintepret(Unsigned, x), sizeof(x)*2)
+@deprecate num2hex(n::Integer) = hex(n, sizeof(n)*2)
```
Comment (tkelman) [inline base/deprecated.jl]: at-deprecate doesn't need the =, right?

Comment (simonbyrne) [inline base/deprecated.jl]: Good point.

**Comment by rfourquet** [comment]
Instead of getting rid of those functions, I would prefer to transfer the functionality somewhere else. I didn't know them until few days ago, but I find `num2hex` in particular useful, as an alternative to `bits` which is harder to read (because counting a lot of 0s and 1s on the screen is difficult. So I propose as the path forward:
1. Fix the bugs in those functions (#22039)
2. deprecate, but incorporate `num2hex` into `bits` via an additional argument (`bits(x, hex=true)`), and `hex2num` maybe in `parse` (cf. e.g. https://github.com/JuliaLang/julia/issues/14418#issuecomment-164956076).

**Comment by ararslan** [comment]
I agree that the functionality should still be easily accessible. Having to do `reinterpret(Float64, parse(UInt64, s, 16))`, for example, is a little unfortunate.

**Comment by simonbyrne** [comment]
> I agree that the functionality should still be easily accessible. Having to do `reinterpret(Float64, parse(UInt64, s, 16))`, for example, is a little unfortunate.

I disagree: it's not clear exactly what the use case of this function is, and as discussed in #22031 it seems to encourage poor practices.

**Comment by rfourquet** [comment]
`hex2num` acts as the opposite of `bits`, so if one is available in `Base`, the other should also be available IMHO. As Julia is a language where we can play easily with bits (and even create primitive bit types!), I feel that those functions are a nice fit in the base library. Everyone seems to agree that the names should be changed though.

**Comment by joa-quim** [comment]
> Everyone seems to agree that the names should be changed though.

I do not. Those are the names used in Matlab (not by coincidence I guess) and I feel that a change will only bring pain to Matlab users for no real gain.

**Comment by tkelman** [comment]
matlab doesn't have dispatch, so a lot of its API naming choices aren't that useful in terms of what we should do (`str2num` anyone?)

**Comment by stevengj** [comment]
@rfourquet, if you just want to read the hex format of a floating-point number, why not use `reinterpret`?
```
julia> reinterpret(UInt64, 3.7)
0x400d99999999999a
```
`bits` is different because there is no native Julia type that prints as binary.

**Comment by rfourquet** [comment]
Yes you are right. Probably when I like to see the bits, in hex or binary form, I think "bits" rather than "reinterpret" in my head ;-) Still, `reinterpret` needs you to specify an unsigned type with the correct width.

**Comment by stevengj** [comment]
We could always add a `reinterpret(Unsigned, x)` method that automatically infers the width.

**Comment by rfourquet** [comment]
That's [exactly](https://github.com/JuliaLang/julia/pull/22203/files#diff-046de3f16395beaddc054d24d670c85fR633) how I implemented `bits(x, hex)` in #22203 ! Still, I find it convenient. And it can be needed also programatically (not only to see it printed on screen).

**Comment by stevengj** [comment]
The question remains, what real application needs a hex form of a Float number as an actual string object that isn't served at least as well by `reinterpret`?  

As I mentioned in #22031, the only usage of `hex2num`/`num2hex` that I found in *any* package would be *greatly* improved by switching to `reinterpret`.

**Comment by rfourquet** [comment]
I'm all for deprecating `num2hex`, but I fill it doesn't overload `Base` too much to add an optional argument to `bits` to have the number displayed in hex instead of binary.

**Comment by stevengj** [comment]
I don't see the value of complicating `bits` in this way, vs. just recommending `reinterpret(Unsigned, x)`.

**Comment by stevengj** [comment]
Adding 1.0 milestone since we should really get a decision on this.

**Comment by JeffBezanson** [comment]
Triage is in favor. Let's rebase and merge.

**Comment by simonbyrne** [comment]
done.

File: base/deprecated.jl [hunk 2]
```diff
@@ -1645,6 +1646,22 @@ end
...
+    if length(s) <= 4
+        return reinterpret(Float16, parse(UInt16, s, 16))
+    end
+    if length(s) <= 8
+        return reinterpret(Float32, parse(UInt32, s, 16))
+    end
+    return reinterpret(Float64, parse(UInt64, s, 16))
+end
+
+@deprecate num2hex(x::Union{Float16,Float32,Float64}) hex(reintepret(Unsigned, x), sizeof(x)*2)
```
Comment (fredrikekre) [inline base/deprecated.jl]: `reintepret` -> `reinterpret`

**Comment by ViralBShah** [comment]
This needs an entry in NEWS.

**Pull request merged**